hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/producer/kafka/MultiDatacenterMessageProducer.java

```diff
@@ -0,0 +1,166 @@
+    }
+
+    @Override
+    public void send(Message message, CachedTopic cachedTopic, PublishingCallback callback) {
+        var producerRecord = messageConverter.convertToProducerRecord(message, cachedTopic.getKafkaTopics().getPrimary().name());
+
+        Optional<KafkaMessageSender<byte[], byte[]>> remoteSender = getRemoteSender(cachedTopic);
+
+        final SendCallback sendCallback = remoteSender.isPresent()
+                ? SendCallback.withFallback(callback)
+                : SendCallback.withoutFallback(callback);
+
+        fallbackScheduler.schedule(() -> {
+            if (!sendCallback.sent.get() && remoteSender.isPresent()) {
+                send(remoteSender.get(),
+                        producerRecord,
+                        sendCallback,
+                        cachedTopic,
+                        message);
+            }
+        }, speculativeSendDelay.toMillis(), TimeUnit.MILLISECONDS);
+
+        send(kafkaMessageSenders.get(cachedTopic.getTopic()), producerRecord, sendCallback, cachedTopic, message);
+    }
+
+    private void send(KafkaMessageSender<byte[], byte[]> sender,
+                      ProducerRecord<byte[], byte[]> producerRecord,
+                      SendCallback callback,
+                      CachedTopic cachedTopic,
+                      Message message) {
+        String datacenter = sender.getDatacenter();
+        try {
+            sender.send(producerRecord, cachedTopic, message, new DCAwareCallback(
+                    message,
+                    cachedTopic,
+                    datacenter,
+                    callback));
+        } catch (Exception e) {
+            // message didn't get to internal producer buffer and it will not be send to a broker
+            callback.onUnpublished(message, cachedTopic, datacenter, e);
+        }
+    }
+
+    private Optional<KafkaMessageSender<byte[], byte[]>> getRemoteSender(CachedTopic cachedTopic) {
+        return kafkaMessageSenders.getRemote(cachedTopic.getTopic())
+                .stream()
+                .filter(producer -> adminReadinessService.isDatacenterReady(producer.getDatacenter()))
+                .findFirst();
+    }
+
+    private record DCAwareCallback(Message message, CachedTopic cachedTopic, String datacenter,
+                                   SendCallback callback) implements Callback {
+
+        @Override
+            public void onCompletion(RecordMetadata metadata, Exception exception) {
+                if (exception == null) {
+                    callback.onPublished(message, cachedTopic, datacenter);
+                } else {
+                    callback.onUnpublished(message, cachedTopic, datacenter, exception);
+                }
+            }
+        }
+
+    private static class SendCallback {
+
+        private final PublishingCallback callback;
+        private final AtomicBoolean sent = new AtomicBoolean(false);
+        private final AtomicInteger tries;
```

Comment:
Just curious: this is atomic because the `onUnpublished` (where the variable is decremented) can be called concurrently be a local and remote producer ?

Reply:
Yes, that's the reason why it is atomic.